src/vos/vos_dtx.c

```diff
@@ -1096,8 +1098,30 @@ vos_dtx_check_shares(struct dtx_handle *dth, struct vos_dtx_entry_df *dtx,
 		return dtx_inprogress(dtx, 5);
 	}
 
-	if (dth == NULL)
-		return dtx_inprogress(dtx, 6);
+	/* Here, if the obj/key has 'prepared' DTX, but current @dth is NULL,
+	 * then it is the race case between normal IO and rebuild: the normal
+	 * IO creates the obj/key before the rebuild request being handled.
+	 *
+	 * Under such case, we should (partially) commit the normal DTX with
+	 * the shared target (obj/key) to guarantee the rebuild can go ahead.
+	 */
```

Comment:
I don't quite understand this, you mean we'll do DTX commit here without leader making the DTX commit decision?

Reply:
Here the object/key should have already been there before. But because old server corruption, we have not rebuilt the object/key in time yet, then the normal IO request came in and created the object/key. But before the DTX for such IO request is committed, the rebuild request arrives and wants to rebuild such object/key, then we should let the rebuild logic mark such object/key as committed instead of waiting for the DTX leader to do that.